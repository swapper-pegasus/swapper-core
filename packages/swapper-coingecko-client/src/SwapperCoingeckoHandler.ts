import {
  ErrorResponse,
  HttpClientError,
  Response,
  ResponseHandler,
  ErrorCoingecko
} from './http/HttpClientInterface'

export default class SwapperCoingeckoHandler implements ResponseHandler {
  handle<T> (response: Promise<Response<T>>): Promise<T> {
    return response.then(
      (res) => this.handleSuccessResponse<T>(res),
      (error) => this.handleErrorResponse(error)
    )
  }

  private handleSuccessResponse<T> (response: Response<T>): T {
    return response.data
  }

  private handleErrorResponse (
    error: HttpClientError<ErrorResponse<ErrorCoingecko>>
  ): never {
    if (error.response?.data?.error) {
      throw new Error(error.response?.data?.error?.message)
    }
    throw new Error(error.message)
  }
}
