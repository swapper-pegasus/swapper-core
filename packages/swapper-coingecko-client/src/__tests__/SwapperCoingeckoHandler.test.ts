/* eslint-disable promise/param-names */
import SwapperCoingeckoHandler from '../SwapperCoingeckoHandler'
import {
  Response
} from '../http/HttpClientInterface'
const handler = new SwapperCoingeckoHandler()

describe('SwapperCoingeckoHandler', () => {
  it('success', async () => {
    const response: Response<string> = { data: 'Success', headers: {} }
    const param: Promise<Response<string>> = new Promise((resolve) => resolve(response))
    const result = await handler.handle(param)
    expect(result).toMatchSnapshot()
  })
  describe('error', () => {
    it('error coingecko', async () => {
      const error = {
        response: {
          data: { error: 'custom error' },
          status: 500,
          statusText: 'Internal server error',
          headers: { header: 'my-header' }
        }
      }
      const param: Promise<Response<string>> = new Promise((_, reject) => reject(error))
      try {
        await handler.handle(param)
      } catch (e) {
        expect(e).toMatchSnapshot()
      }
    })
    it('error without data', async () => {
      const error = {
        response: {
          status: 500,
          statusText: 'Internal server error',
          headers: { header: 'my-header' }
        }
      }
      const param: Promise<Response<string>> = new Promise((_, reject) => reject(error))
      try {
        await handler.handle(param)
      } catch (e) {
        expect(e).toMatchSnapshot()
      }
    })
    it('error without response', async () => {
      const error = {}
      const param: Promise<Response<string>> = new Promise((_, reject) => reject(error))
      try {
        await handler.handle(param)
      } catch (e) {
        expect(e).toMatchSnapshot()
      }
    })
  })
})
