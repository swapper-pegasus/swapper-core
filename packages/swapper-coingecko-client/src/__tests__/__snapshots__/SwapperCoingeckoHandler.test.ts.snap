// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwapperCoingeckoHandler error error coingecko 1`] = `[Error: custom error]`;

exports[`SwapperCoingeckoHandler error error without data 1`] = `[Error]`;

exports[`SwapperCoingeckoHandler error error without response 1`] = `[Error]`;

exports[`SwapperCoingeckoHandler success 1`] = `"Success"`;
